3scale/3scale-operator
# THREESCALE-8754 Enhance replica reconciliation

### what
Implements https://issues.redhat.com/browse/THREESCALE-8754

* Remove adding replica values to the CR
* Reconciliation (implemented by the mutator) only added when the `replicas` field is set

Components
- [x] backend listener
- [x] backend worker
- [x] backend cron
- [x] apicast staging
- [x] apicast production
- [x] system app
- [x] system sidekiq
- [x] zync
- [x] zync que

### Verification steps

run env

```
make install
make run
```
Deploy APIManager CR with no replica definition
```yaml
k apply -f -<<EOF
---
apiVersion: v1
kind: Secret
metadata:
  creationTimestamp: null
  name: aws-auth
stringData:
  AWS_ACCESS_KEY_ID: somekey
  AWS_SECRET_ACCESS_KEY: something
  AWS_BUCKET: example.com
  AWS_REGION: us-east-1
type: Opaque
---
apiVersion: apps.3scale.net/v1alpha1
kind: APIManager
metadata:
  name: apimanager1
spec:
  wildcardDomain: example.com
  resourceRequirementsEnabled: false
  system:
    fileStorage:
      simpleStorageService:
        configurationSecretRef:
          name: aws-auth
EOF
```

Check that the APIManager  CR does not have replicas in the CR

```yaml
k get apimanager.apps.3scale.net/apimanager1 -o yaml
apiVersion: apps.3scale.net/v1alpha1
kind: APIManager
metadata:
  annotations:
    apps.3scale.net/apimanager-threescale-version: "2.13"
    apps.3scale.net/threescale-operator-version: 0.10.0
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"apps.3scale.net/v1alpha1","kind":"APIManager","metadata":{"annotations":{},"name":"apimanager1","namespace":"3scale-operator-test"},"spec":{"resourceRequirementsEnabled":false,"system":{"fileStorage":{"simpleStorageService":{"configurationSecretRef":{"name":"aws-auth"}}}},"wildcardDomain":"3scale-operator-test.apps.dev-eng-ocp4-operator.dev.3sca.net"}}
  creationTimestamp: "2022-09-29T09:34:19Z"
  generation: 2
  name: apimanager1
  namespace: 3scale-operator-test
  resourceVersion: "9968928"
  uid: 14854404-5511-43fc-bfbb-181155e50aa3
spec:
  apicast:
    managementAPI: status
    openSSLVerify: false
    productionSpec: {}
    registryURL: http://apicast-staging:8090/policies
    responseCodes: true
    stagingSpec: {}
  appLabel: 3scale-api-management
  backend:
    cronSpec: {}
    listenerSpec: {}
    workerSpec: {}
  imageStreamTagImportInsecure: false
  resourceRequirementsEnabled: false
  system:
    appSpec: {}
    fileStorage:
      simpleStorageService:
        configurationSecretRef:
          name: aws-auth
    sidekiqSpec: {}
    sphinxSpec: {}
  tenantName: 3scale
  wildcardDomain: 3scale-operator-test.apps.dev-eng-ocp4-operator.dev.3sca.net
  zync:
    appSpec: {}
    queSpec: {}
```

Wait for deployment to be available
```
oc wait --for=condition=available apimanager/apimanager1 --timeout=-1s
apimanager.apps.3scale.net/apimanager1 condition met
```
Let's pick one subcomponent (it should be the same for any one where you can specify replicas): apicast-staging

Check the number of pods is 1 for the apicast staging
```
k get dc apicast-staging
NAME              REVISION   DESIRED   CURRENT   TRIGGERED BY
apicast-staging   1          1         1         config,image(amp-apicast:2.13)
```

Scale to 3 via the deploymentconfig

```
k scale --replicas=3 deploymentconfig apicast-staging
Warning: extensions/v1beta1 Scale is deprecated in v1.2+, unavailable in v1.16+
deploymentconfig.apps.openshift.io/apicast-staging scaled
```
Check the deployment has 3 replicas

```
k get pods -l deploymentConfig=apicast-staging
NAME                      READY   STATUS    RESTARTS   AGE
apicast-staging-1-7ds4r   1/1     Running   0          14m
apicast-staging-1-l8l95   1/1     Running   0          106s
apicast-staging-1-lfstj   1/1     Running   0          106s
```

Add `replicas` to the APIManager CR with a value of 2

```
k patch apimanager apimanager1 --type merge --patch '{"spec":{"apicast": {"stagingSpec": {"replicas": 2}}}}'
apimanager.apps.3scale.net/apimanager1 patched
```
Check the deployment has 2 replicas
```
k get pods -l deploymentConfig=apicast-staging
NAME                      READY   STATUS    RESTARTS   AGE
apicast-staging-1-7ds4r   1/1     Running   0          16m
apicast-staging-1-lfstj   1/1     Running   0          4m21s
```
As replicas are set explicitly in the CR, changes in the deployment will be reverted back. If we try to scale to 3 via the deployment

```
k scale --replicas=3 deploymentconfig apicast-staging
```
The number of replicas will still be 2
```
k get pods -l deploymentConfig=apicast-staging
NAME                      READY   STATUS    RESTARTS   AGE
apicast-staging-1-7ds4r   1/1     Running   0          17m
apicast-staging-1-lfstj   1/1     Running   0          5m32s
```
